fhir_pyrate/ahoy.py

```diff
@@ -139,8 +142,25 @@ def refresh_token(self, token: str = None) -> None:
             self.token = response.text
             self.session.headers.update({"Authorization": f"Bearer {self.token}"})
             self.auth_time = datetime.datetime.now()
-        else:
-            logger.warning(
-                "The token cannot be refreshed because neither a valid token nor a refresh url has "
-                "been provided."
+        elif self.auth_type is not None and self.auth_method is not None:
+            self._authenticate()
+            self.auth_time = datetime.datetime.now()
+
+    def _refresh_hook(
+        self, response: requests.Response, *args: Any, **kwargs: Any
+    ) -> Optional[requests.Response]:
+        if self.login_reattempted_times >= self.max_login_attempts:
+            response.raise_for_status()
+        elif response.status_code == requests.codes.unauthorized:
+            self.login_reattempted_times += 1
+            self.token = None
+            self.refresh_session()
```

Comment:
Currently, tokens will never be refreshed using the `refresh_url`, since this is the only point of invocation and the token will be erased beforehand.